crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -310,152 +205,225 @@ impl IndexAllocator {
 
     pub(crate) fn free(&self, index: SlotId) {
         let mut inner = self.0.lock().unwrap();
-        let free_list_index = GlobalFreeListIndex(inner.free_list.len());
-        inner.free_list.push(index);
-        let module_id = inner.slot_state[index.index()].unwrap_module_id();
-        inner.slot_state[index.index()] = if let Some(id) = module_id {
-            let per_module_list = inner
-                .per_module
-                .entry(id)
-                .or_insert_with(|| Vec::with_capacity(1));
-            let per_module_index = PerModuleFreeListIndex(per_module_list.len());
-            per_module_list.push(index);
-            SlotState::Free(FreeSlotState::Affinity {
-                module: id,
-                free_list_index,
-                per_module_index,
-            })
-        } else {
-            SlotState::Free(FreeSlotState::NoAffinity { free_list_index })
+        let inner = &mut *inner;
+        let module = match inner.slot_state[index.index()] {
+            SlotState::Used(module) => module,
+            _ => unreachable!(),
         };
+
+        // Bump the number of warm slots since this slot is now considered
+        // previously used. Afterwards append it to the linked list of all
+        // unused and warm slots.
+        inner.unused_warm_slots += 1;
+        let unused = inner
+            .warm
+            .append(index, &mut inner.slot_state, |s| &mut s.unused);
+
+        let affine = match module {
+            // If this slot is affine to a particular module then append this
+            // index to the linked list for the affine module. Otherwise insert
+            // a new one-element linked list.
+            Some(module) => match inner.module_affine.entry(module) {
+                Entry::Occupied(mut e) => e
+                    .get_mut()
+                    .append(index, &mut inner.slot_state, |s| &mut s.affine),
+                Entry::Vacant(v) => {
+                    v.insert(List::new(index));
+                    Link::default()
+                }
+            },
+
+            // If this slot has no affinity then the affine link is empty.
+            None => Link::default(),
+        };
+
+        inner.slot_state[index.index()] = SlotState::UnusedWarm(Unused {
+            affinity: module,
+            affine,
+            unused,
+        });
     }
 
     /// For testing only, we want to be able to assert what is on the
     /// single freelist, for the policies that keep just one.
     #[cfg(test)]
     pub(crate) fn testing_freelist(&self) -> Vec<SlotId> {
         let inner = self.0.lock().unwrap();
-        inner.free_list.clone()
+        inner.warm.iter(&inner.slot_state, |s| &s.unused).collect()
```

Comment:
Can we debug-assert here that `unused_warm` is correct, since we're collecting the linked list into a `Vec` anyway?